### PR TITLE
magento/magento2: Fixes for the schema cache.xsd

### LINKED ISSUE #26224: Cache type without "instance" causes exception when disabling the module through "Cache Management" in the backend
### Preconditions (*)
1. Magento 2.3.3 CE

### Steps to reproduce (*)

1. Create a cache type using "cache.xml" without the "instance" argument
```xml
<config .../>
    <type name="exceptional_cache_type">
        <label>Exceptional Cache Type</label>
        <description>An exceptional cache type using a quantum computer</description>
    </type>
</config>
```
2. Go to "System > Cache Management" in the backend
3. Check the checkbox of the cache type "Exceptional Cache Type"
4. Select "Disable" and click "Submit"

### Expected result (*)
I'd expect the cache type to be successfully disabled. 

According to the XSD ("lib/internal/Magento/Framework/Cache/etc/cache.xsd") the "instance" argument is optional and not required.

### Actual result (*)
It throws an exception because it's trying to invoke `->clean(...)` on a null variable. 

```
Fatal error: Uncaught Error: Call to a member function clean() on null in /lib/internal/Magento/Framework/App/Cache/TypeList.php:194 Stack trace:
#0 /app/code/Magento/Backend/Controller/Adminhtml/Cache/MassDisable.php(62): Magento\Framework\App\Cache\TypeList->cleanType('invalid_cache_t...')
#1 /app/code/Magento/Backend/Controller/Adminhtml/Cache/MassDisable.php(41): Magento\Backend\Controller\Adminhtml\Cache\MassDisable->disableCache()
#2 /generated/code/Magento/Backend/Controller/Adminhtml/Cache/MassDisable/Interceptor.php(24): Magento\Backend\Controller\Adminhtml\Cache\MassDisable->execute()
#3 /lib/internal/Magento/Framework/App/Action/Action.php(107): Magento\Backend\Controller\Adminhtml\Cache\MassDisable\Interceptor->execute()
```

### PR DESCRIPTION
**Pull request by andrewbess**
### Description (*)

This PR fixes the schema "cache.xsd" from Magento\Framework\Cache.

1. The attribute "instance" has changed on required because if a developer won't be to use it new cache type will not be created;
2. The parameters "label" and "description" have changed on required because of the class "Magento\Framework\App\Cache\TypeList" gets theirs without checking for existing. Please take a look at [this code](https://github.com/magento/magento2/blob/cde4f67d8e02ad7ffe1de5051010bdce71fdabfd/lib/internal/Magento/Framework/App/Cache/TypeList.php#L124);
3. Checking on unique of the type name has added;
4. The configuration reader class has fixed for more informative during to run CLI commands;
5. The class SchemaLocator has been fixed to check the cache configuration during configuration merging;
6. The unnecessary checking has removed from the class TypeList;
7. Unit tests have been covered.

### Fixed Issues (if relevant)

Fixes: https://github.com/magento/magento2/issues/26224

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Review by swnsma — CHANGES_REQUESTED** [review]
Hi @andrewbess!
Thank you for your contribution!
Could you please add unit tests to check new schema validation?
Maybe, we also should refactor the class \Magento\Framework\App\Cache\TypeList as 'instance' will become a required parameter? We may simplify some checks here.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Please see following review details: https://github.com/magento/magento2/pull/26933#pullrequestreview-362147607

**Comment by andrewbess** [comment]
> Hi @andrewbess!
> Thank you for your contribution!
> Could you please add unit tests to check new schema validation?
> Maybe, we also should refactor the class \Magento\Framework\App\Cache\TypeList as 'instance' will become a required parameter? We may simplify some checks here.

Hello @swnsma 
Thank you for review.
I added the changes.
There are additional changes for schema, tests, additional changes for TypeList in the PR.
Also, I have changed description of PR.
Please recheck.
Thank you in advance.

File: lib/internal/Magento/Framework/Cache/Config/Reader.php [hunk 1]
```diff
@@ -45,4 +50,25 @@ public function __construct(
...
+     *
+     * @param string|null $scope
+     * @return array
+     * @throws \Magento\Framework\Exception\LocalizedException
+     */
+    public function read($scope = null)
+    {
+        try {
+            return parent::read($scope);
+        } catch (\Magento\Framework\Exception\LocalizedException $e) {
```
Comment (rogyar) [inline lib/internal/Magento/Framework/Cache/Config/Reader.php]: May I ask you why do we need to transform this localized exception and throw once again?

Thank you

File: lib/internal/Magento/Framework/Cache/Test/Unit/XsdTest.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+namespace Magento\Framework\Cache\Test\Unit;
+
+class XsdTest extends \PHPUnit\Framework\TestCase
```
Comment (swnsma) [inline lib/internal/Magento/Framework/Cache/Test/Unit/XsdTest.php]: Hi @andrewbess,

I have discovered that we already have bunch of tests which test *.xsd files.
They are located in Magento\Test\Integrity namespace.
Could I ask you to move your your tests under this namespace to keep consistency in tests structure?
Please check \Magento\Test\Integrity\Magento\Widget\WidgetConfigTest for example.

**Review by swnsma — CHANGES_REQUESTED** [review]
Please check my comment. 
Would you be so kind to check strict types declaration, type hints and add 'declare(strict_types=1);' to changed files.
Thank you!

File: lib/internal/Magento/Framework/Cache/Config/Reader.php (hunk 1 above)
Comment (andrewbess) [reply lib/internal/Magento/Framework/Cache/Config/Reader.php]: Hello @rogyar.
Thank you for your remark.
Previously, I have added the render-method that prepares an error message to output error messages correctly.
But it caused other problems.
I have removed new methods.
Please recheck.
Thank you in advance.

File: lib/internal/Magento/Framework/Cache/Test/Unit/XsdTest.php (hunk 2 above)
Comment (andrewbess) [reply lib/internal/Magento/Framework/Cache/Test/Unit/XsdTest.php]: Hello @swnsma.
I have moved tests by the suggested path.
Also, I added declaration strict types where it is required.
Please recheck.
Thank you in advance.

File: dev/tests/static/testsuite/Magento/Test/Integrity/Magento/Framework/Cache/ConfigTest.php [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    /**
+     * Tests invalid configurations
+     *
+     * @param string $xmlString
+     * @param array $expectedError
+     * @throws \Magento\Framework\Exception\NotFoundException
+     * @dataProvider schemaCorrectlyIdentifiesInvalidXmlDataProvider
+     */
+    public function testSchemaCorrectlyIdentifiesInvalidXml(
+        $xmlString,
```
Comment (swnsma) [inline dev/tests/static/testsuite/Magento/Test/Integrity/Magento/Framework/Cache/ConfigTest.php]: Please add strict type 'string'.

File: dev/tests/static/testsuite/Magento/Test/Integrity/Magento/Framework/Cache/ConfigTest.php [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+    /**
+     * @var \Magento\Framework\Config\Dom\UrnResolver
+     */
+    protected $urnResolver;
+
+    /**
+     * @var \Magento\Framework\TestFramework\Unit\Utility\XsdValidator
+     */
+    protected $_xsdValidator;
```
Comment (swnsma) [inline dev/tests/static/testsuite/Magento/Test/Integrity/Magento/Framework/Cache/ConfigTest.php]: Please
**protected -> private**
**$_xsdValidator -> $xsdValidator**

File: dev/tests/static/testsuite/Magento/Test/Integrity/Magento/Framework/Cache/ConfigTest.php [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    /**
+     * Path to xsd schema file
+     * @var string
+     */
+    protected $_xsdSchema;
+
+    /**
+     * @var \Magento\Framework\Config\Dom\UrnResolver
+     */
+    protected $urnResolver;
```
Comment (swnsma) [inline dev/tests/static/testsuite/Magento/Test/Integrity/Magento/Framework/Cache/ConfigTest.php]: Please
**protected -> private**

File: dev/tests/static/testsuite/Magento/Test/Integrity/Magento/Framework/Cache/ConfigTest.php [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+/**
+ * Unit test of the cache configuration
+ */
+class ConfigTest extends \PHPUnit\Framework\TestCase
+{
+    /**
+     * Path to xsd schema file
+     * @var string
+     */
+    protected $_xsdSchema;
```
Comment (swnsma) [inline dev/tests/static/testsuite/Magento/Test/Integrity/Magento/Framework/Cache/ConfigTest.php]: Please
**protected -> private**
**$_xsdSchema -> $xsdSchema**

File: lib/internal/Magento/Framework/Cache/Exception/WrongConfigurationException.php [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+ */
+
+namespace Magento\Framework\Cache\Exception;
+
+/**
+ * Exception for situation when cache configuration is wrong
+ *
+ * @api
+ */
+class WrongConfigurationException extends \Exception
```
Comment (swnsma) [inline lib/internal/Magento/Framework/Cache/Exception/WrongConfigurationException.php]: Sorry, I can not find place where this kind of exception is used...
Could you advice?

File: lib/internal/Magento/Framework/App/Test/Unit/Cache/TypeListTest/Type.php [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace Magento\Framework\App\Test\Unit\Cache\TypeListTest;
+
+use Magento\Framework\Cache\Frontend\Decorator\TagScope;
+use Magento\Framework\App\Test\Unit\Cache\TypeListTest;
+
+/**
+ * Cache type "Type"
+ */
+class Type extends TagScope
```
Comment (swnsma) [inline lib/internal/Magento/Framework/App/Test/Unit/Cache/TypeListTest/Type.php]: I am not sure if we need class to only handle constants.
Can we just use private constants in places where they need to be used?

**Review by swnsma — CHANGES_REQUESTED** [review]
Hi @andrewbess!
Thank you for your contribution!
Could you please check my notices?

**Comment by andrewbess** [comment]
Closed in favor of https://github.com/magento/magento2/pull/27307

**Pull request closed without merging**